src/ImageSharp/Formats/Png/Zlib/Crc32.cs

```diff
@@ -4,151 +4,206 @@
-            0xCDD70693, 0x54DE5729, 0x23D967BF, 0xB3667A2E, 0xC4614AB8,
-            0x5D681B02, 0x2A6F2B94, 0xB40BBE37, 0xC30C8EA1, 0x5A05DF1B,
-            0x2D02EF8D
-        };
+        /// <param name="buffer">The readonly span of bytes.</param>
+        /// <returns>The <see cref="uint"/>.</returns>
+        [MethodImpl(InliningOptions.ShortMethod)]
+        public static uint Calculate(ReadOnlySpan<byte> buffer)
+            => Calculate(SeedValue, buffer);
 
         /// <summary>
-        /// The data checksum so far.
+        /// Calculates the CRC checksum with the bytes taken from the span and seed.
         /// </summary>
-        private uint crc;
-
-        /// <inheritdoc/>
-        public long Value
+        /// <param name="crc">The input CRC value.</param>
+        /// <param name="buffer">The readonly span of bytes.</param>
+        /// <returns>The <see cref="uint"/>.</returns>
+        [MethodImpl(InliningOptions.ShortMethod)]
+        public static uint Calculate(uint crc, ReadOnlySpan<byte> buffer)
         {
-            [MethodImpl(MethodImplOptions.AggressiveInlining)]
-            get => this.crc;
+            if (buffer.IsEmpty)
+            {
+                return SeedValue;
+            }
 
-            [MethodImpl(MethodImplOptions.AggressiveInlining)]
-            set => this.crc = (uint)value;
+#if SUPPORTS_RUNTIME_INTRINSICS
+            if (Sse41.IsSupported && Pclmulqdq.IsSupported && buffer.Length >= MinBufferSize)
+            {
+                return ~CalculateSse(~crc, buffer);
+            }
+            else
+            {
+                return ~CalculateScalar(~crc, buffer);
+            }
+#else
+            return ~CalculateScalar(~crc, buffer);
+#endif
         }
 
-        /// <inheritdoc/>
-        [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        public void Reset()
+#if SUPPORTS_RUNTIME_INTRINSICS
+        // Based on https://github.com/chromium/chromium/blob/master/third_party/zlib/crc32_simd.c
+        [MethodImpl(InliningOptions.HotPath | InliningOptions.ShortMethod)]
+        private static unsafe uint CalculateSse(uint crc, ReadOnlySpan<byte> buffer)
         {
-            this.crc = 0;
-        }
+            int chunksize = buffer.Length & ~ChunksizeMask;
+            int length = chunksize;
 
-        /// <summary>
-        /// Updates the checksum with the given value.
-        /// </summary>
-        /// <param name="value">The byte is taken as the lower 8 bits of value.</param>
-        [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        public void Update(int value)
-        {
-            this.crc ^= CrcSeed;
-            this.crc = CrcTable[(this.crc ^ value) & 0xFF] ^ (this.crc >> 8);
-            this.crc ^= CrcSeed;
+            fixed (byte* bufferPtr = &buffer[0])
+            fixed (ulong* k1k2Ptr = &k1k2[0])
+            fixed (ulong* k3k4Ptr = &k3k4[0])
+            fixed (ulong* k5k0Ptr = &k5k0[0])
+            fixed (ulong* polyPtr = &poly[0])
```

Comment:
This is another place where the ROS<byte> trick would be of benefit. Instead of 4 separate unmanaged pointers, you could have a single pointer to a block containing all 4 vector values.

Reply:
I've updated the code to use a static readonly array as I wasn't sure the ROS optimizations apply for anything other than byte?

Reply:
Yep, the compiler won't optimize anything bigger than byte because of endianness concerns. Since this code is processor-specific, you can outsmart the compiler by breaking the ulongs up into bytes (and reversing their byte order, ofc)

Reply:
I'll leave that for now since the massive hashing improvements have actually made less of an impact than I'd hoped. I'm going to profile now and find other soft targets.